core/src/main/java/com/google/cloud/sql/TokenSourceFactory.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.auth.oauth2.OAuth2Credentials;
+
+/**
+ * Factory for creating a token source from {@link Credential}s for IAM Authentication.
+ */
+public interface TokenSourceFactory {
+
+  /**
+   * Name of system property that can specify an alternative credential factory.
+   */
+  String TOKEN_SOURCE_FACTORY_PROPERTY = "cloudSql.socketFactory.TokenSourceFactory";
```

Comment:
We already have a credential socket factory right? Can we get a token source from the credentials?

Reply:
I guess what I'm saying is we already have a property for setting the user credentials, can we reuse the user credentials instead of trying to ask for the token source from somewhere else?

Reply:
So the Credential Factory returns credentials that are wrapped in a `HttpCredentialsAdapter` and there's no way to get credentials back from that, because they're private https://github.com/googleapis/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/http/HttpCredentialsAdapter.java#L64

We could change the Credential Factory to return a `GoogleCredentials` object and then convert that to the `HttpCredentialsAdapter` later, but I'm hesitant to change the existing interface

Reply:
That's a DPE owned library - can we PR a getter so we can have access to it?

Reply:
https://github.com/googleapis/google-auth-library-java/pull/658